src/api/workspaceManager.ts

```diff
@@ -369,202 +171,167 @@ export default class WorkspaceManager {
+     * Refreshes current active course workspace by first making sure that the `.tmc` folder is at
+     * the top and then lists all that course's open exercises in alphanumeric order.
      */
-    private _verifyCourseWorkspaceRootFolderAndFileExists(): void {
-        const rootFileFolder = path.join(this._resources.workspaceFolderPath, ".tmc");
-        const pathToRootFile = path.join(rootFileFolder, WORKSPACE_ROOT_FILE);
-        if (!fs.existsSync(pathToRootFile)) {
-            Logger.log(`Creating ${pathToRootFile}`);
-            fs.mkdirSync(rootFileFolder, { recursive: true });
-            fs.writeFileSync(pathToRootFile, WORKSPACE_ROOT_FILE_TEXT, { encoding: "utf-8" });
-        } else if (
-            fs.readFileSync(pathToRootFile, { encoding: "utf-8" }) !== WORKSPACE_ROOT_FILE_TEXT
+    private async _refreshActiveCourseWorkspace(): Promise<Result<void, Error>> {
+        const workspaceName = this._getActiveCourseWorkspace();
+        const workspaceFolders = vscode.workspace.workspaceFolders;
+        if (!workspaceName || !workspaceFolders) {
+            Logger.warn("Attempted refresh for a non-course workspace.");
+            return Ok.EMPTY;
+        }
+
+        const rootFolder = this._resources.workspaceRootFolder;
+        const openExercises = this._exercises
+            .filter((x) => x.status === ExerciseStatus.Open)
+            .sort((a, b) => a.exerciseSlug.localeCompare(b.exerciseSlug))
+            .map((x) => ({ uri: x.uri }));
+        const correctStructure = [{ uri: rootFolder }, ...openExercises];
+        if (
+            _.zip(correctStructure, workspaceFolders).every(
+                ([a, b]) => a?.uri.fsPath === b?.uri.fsPath,
+            )
         ) {
-            Logger.log(`Rewriting ${WORKSPACE_ROOT_FILE_TEXT} at ${pathToRootFile}`);
-            fs.writeFileSync(pathToRootFile, WORKSPACE_ROOT_FILE_TEXT, { encoding: "utf-8" });
+            Logger.debug("Workspace refresh was a no-op.");
+            return Ok.EMPTY;
+        }
+
+        Logger.log("Refreshing workspace structure in the current workspace");
+        if (workspaceFolders[0]?.name !== WORKSPACE_ROOT_FOLDER_NAME) {
+            Logger.warn("Fixing incorrect root folder. This may restart the extension.");
+        }
+
+        const deleteCount = workspaceFolders.length;
+        Logger.debug(`Replacing ${deleteCount} workspace folders with ${correctStructure.length}`);
+        const success = vscode.workspace.updateWorkspaceFolders(
+            0,
+            deleteCount,
+            ...correctStructure,
+        );
+        if (!success) {
+            Logger.error("Replace operation failed.");
+            Logger.debug("Failed with folders:", ...correctStructure);
+        }
+
+        return success ? Ok.EMPTY : Err(new Error("Failed to refresh active workspace."));
+    }
+
+    private _onDidChangeWorkspaceFolders(e: vscode.WorkspaceFoldersChangeEvent): void {
```

Comment:
For some reason this doesn't fire if I manually remove folders from the workspace. That is: Right click a root folder and "Remove folder from Workspace"